stream_alert/alert_processor/outputs/pagerduty.py

```diff
@@ -610,70 +623,131 @@ def run(self, alert, descriptor):
 
-        # FIXME (derek.wang), see above
-        #   At this point, both the incident and the relevant alert event have been successfully
-        #   created. Any further work that fails the dispatch call will cause the alert to retry
-        #   and redundantly create more incidents and alert events.
-        #   Therefore, the hack is to simply let further failures go by always returning True.
-        #   The tradeoff is that incidents can be created on pagerduty in an incomplete state,
-        #   but this is easier to manage than StreamAlert redundantly creating hundreds (or more!)
-        #   redundant alerts.
-        stable = True
-
-        # Merge the incident with the event, so we can have a rich context incident
-        # assigned to a specific person, which the PagerDuty REST API v2 does not allow
-        merged_incident = self._merge_event_into_incident(incident, event)
-        if not merged_incident:
+        # Create an incident to house the alert
+        incident = self._update_base_incident(event, alert, publication, rule_context)
+        if not incident:
             LOGGER.error(
-                '[%s] Failed to merge alert [%s] into [%s]',
-                self._output.__service__,
-                event.get('dedup_key'),
-                incident_id
+                '[%s] Failed to update container incident for event',
+                self._output.__service__
             )
-            stable = False
+            return False
 
-        if merged_incident:
-            note = self._add_incident_note(incident, publication, rule_context)
-            if not note:
-                LOGGER.error(
-                    '[%s] Failed to add note to incident (%s)',
-                    self._output.__service__,
-                    incident_id
+        incident_id = incident.get('id', False)
+        if not incident_id:
+            LOGGER.error(
+                '[%s] Incident is missing "id"??',
+                self._output.__service__
+            )
+            return False
+
+        # At this point, both the incident and the relevant alert event have been successfully
+        # created.
+        #
+        # All of the code above this line is considered idempotent and can be called repeatedly
+        # without adverse side effects. Code BELOW this line is neither atomic nor idempotent, so
+        # we will not retry if any of the below code fails. Instead, we log an error and make a
+        # best-effort attempt to attach an error note to the PagerDuty incident, signalling that
+        # it was not setup properly.
+        #
+        # In the fateful event the alert gets stuck ANYWAY, the easiest solution is to destroy the
+        # associated record on the DynamoDB table.
+        errors = []
+
+        # Add responder requests
+        responders = rule_context.get('responders', [])
+        if responders and not isinstance(responders, list):
+            responders = [responders]
+
+        if responders:
```

Comment:
are responders optional? meaning should we return False `if not responder` or is that okay?

Reply:
Responders are optional; it's a new feature so I need to maintain reverse compatibility.